### ISSUE TITLE
Result seems not right

### ISSUE THREAD

**Issue by zhouzhenghong-gt** [chenhsuanlin/bundle-adjusting-NeRF#23]
Hello, thanks for your great work!
I ran the barf code on the chair of the blender dataset, but the validation result doesn't seem right on tensorboard, is this normal?
Train:
![MZO$U7N$IG`($FQ`(L(}BOT](https://user-images.githubusercontent.com/50356920/157035837-837e1b25-f322-4f4c-8075-9ef78b56defa.png)

Test:
![loss](https://user-images.githubusercontent.com/50356920/157035872-0d9f2508-b5c4-42af-95e3-373c26e4610a.png)

**Comment by chenhsuanlin** [comment]
Hi @zhouzhenghong-gt, thanks for point this out. You're right that the current validation error is more or less flawed. In the current codebase, the validation error is computed directly between the rendered image at the "given" val pose and the ground-truth image. However, since the entire (reference) coordinate system could be changing, the actual validation rendered image may not be aligned with the ground truth, thus the non-decreasing error.

A better way would indeed be to pre-align the poses with Procrustes analysis (as done in eval mode, see the "evaluation criteria" paragraph in the paper), and I can push an update in the coming days to fix this. However, further test-time optimization would be too computationally prohibitive just for the sake of validation. So even with the upcoming fix, the val loss/PSNR would still not be a perfect metric and should be taken with a grain of salt.

Hope this clarifies! Leaving this issue open for now until the fix is updated.

**Comment by zhouzhenghong-gt** [comment]
Thanks for your reply! It helps a lot.

By the way, I am curious if you have compared the results of nerf--?

**Comment by chenhsuanlin** [comment]
I have not unfortunately, would be an interesting comparison though! It should be fairly easy to incorporate NeRF-- into this BARF codebase, essentially (a) additional intrinsics optimization and (b) no coarse to fine in pose optimization.

**Comment by chenhsuanlin** [comment]
Hi @zhouzhenghong-gt, sorry for the delay -- I have just pushed the fix (see 803291b). The code now validates on the pre-aligned poses with Procrustes analysis, and you should see the validation loss/PSNR and rendered images as below.
Hope this helps!

<img width="696" alt="image" src="https://user-images.githubusercontent.com/9806735/161929426-d5c97c20-7081-49d3-a032-24b684fcf15a.png">. 
![image](https://user-images.githubusercontent.com/9806735/161929911-d9ac0307-286c-4c4c-bd1c-04171d0592cf.png)
